sensor/common/wal/wal.go

```diff
@@ -0,0 +1,78 @@
+package wal
+
+import (
+	"encoding/binary"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	walOnce sync.Once
+	wal     WAL
+)
+
+type WAL interface {
+	Insert(id string, hash uint64) error
+	Delete(id string) error
+	GetMap() map[string]uint64
+}
+
+func OpenWAL() WAL {
+	walOnce.Do(func() {
+		db, err := pebble.Open("/var/cache/stackrox/wal", &pebble.Options{})
+		if err != nil {
+			log.Errorf("could not open WAL for Sensor. This could have adverse performance impacts: %v", err)
+		}
+		wal = &walImpl{
+			db: db,
+		}
+	})
+	return wal
+}
+
+type Entry struct {
+	ID   string
+	Hash int64
+}
+
+type walImpl struct {
+	db *pebble.DB
+}
+
+func (w *walImpl) Insert(id string, hash uint64) error {
+	if w.db == nil {
+		return nil
+	}
+	hashBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(hashBytes, hash)
+	return w.db.Set([]byte(id), hashBytes, &pebble.WriteOptions{
+		Sync: true,
```

Comment:
Why sync? If I am not mistaken, `Insert()` is only ever called when flushing out a batch of elements every 10s. Syncing after every element seems very inefficient and also unneeded. I guess there is _some_ advantage wrt. making sure state remains consistent and you get incremental benefit even if sensor crashes while updating the pebble DB, but.. eh. Doesn't seem to be very meaningful.

That being said, I also don't think pebble is the right tool for the job. You are neither using random inserts nor random lookups. All you need is a batch write and a batch read into memory. Something mimicking [RecordIO](https://news.ycombinator.com/item?id=23589534) might be better suited. Since `id` and the little endian encoding of a uint64 are fixed size, you could even just write out the raw data without any size indicators or delimiters (not saying this is a good idea but it would be simple and way more efficient).

Reply:
Yeah I definitely think not syncing would be fine here and can instead just batch everything. Would need to look at performance of other options